pkg/backup/compaction_dist.go

```diff
@@ -0,0 +1,336 @@
+		corePlacements[i].SQLInstanceID = sqlInstanceIDs[i]
+		corePlacements[i].Core.CompactBackups = &execinfrapb.CompactBackupsSpec{
+			JobID:       int64(jobID),
+			DefaultURI:  details.URI,
+			Destination: details.Destination,
+			Encryption:  details.EncryptionOptions,
+			StartTime:   details.StartTime,
+			EndTime:     details.EndTime,
+			ElideMode:   elideMode,
+			UserProto:   user.EncodeProto(),
+			Spans:       spansToCompact,
+			TargetSize:  targetSize,
+			MaxFiles:    maxFiles,
+		}
+	}
+
+	spanEntryCh := make(chan execinfrapb.RestoreSpanEntry, 1000)
+	var tasks []func(ctx context.Context) error
+	tasks = append(tasks, func(ctx context.Context) error {
+		defer close(spanEntryCh)
+		return genSpan(ctx, spanEntryCh)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		numEntriesPerNode := numEntries / numNodes
+		leftoverEntries := numEntries % numNodes
+		getTargetNumEntries := func(nodeIdx int) int {
+			if nodeIdx < leftoverEntries {
+				// This more evenly distributes the leftover entries across the nodes
+				// after doing integer division to assign the entries to the nodes.
+				return numEntriesPerNode + 1
+			}
+			return numEntriesPerNode
+		}
+		currNode := 0
+		entriesAdded := 0
+		currEntries := roachpb.SpanGroup{}
+		targetNumEntries := getTargetNumEntries(currNode)
+
+		for entry := range spanEntryCh {
+			currEntries.Add(entry.Span)
+			entriesAdded++
+			if entriesAdded == targetNumEntries {
+				corePlacements[currNode].SQLInstanceID = sqlInstanceIDs[currNode]
+				corePlacements[currNode].Core.CompactBackups.AssignedSpans = currEntries.Slice()
+
+				currNode++
+				targetNumEntries = getTargetNumEntries(currNode)
+				currEntries.Clear()
+				entriesAdded = 0
+			}
+			if currNode == numNodes {
+				return nil
+			}
+		}
+		return nil
+	})
+	if err := ctxgroup.GoAndWait(ctx, tasks...); err != nil {
+		return nil, errors.Wrapf(err, "distributing span entries to processors")
+	}
+	return corePlacements, nil
+}
+
+// getSpansToCompact returns all remaining spans the backup manifest that
+// need to be compacted.
+func getSpansToCompact(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	manifest *backuppb.BackupManifest,
+	backupChain []backuppb.BackupManifest,
+	details jobspb.BackupDetails,
+	defaultStore cloud.ExternalStorage,
+	kmsEnv cloud.KMSEnv,
+) (roachpb.Spans, error) {
+	var tables []catalog.TableDescriptor
+	for _, desc := range manifest.Descriptors {
+		catDesc := backupinfo.NewDescriptorForManifest(&desc)
+		if table, ok := catDesc.(catalog.TableDescriptor); ok {
+			tables = append(tables, table)
+		}
+	}
+	backupCodec, err := backupinfo.MakeBackupCodec(backupChain)
+	if err != nil {
+		return nil, err
+	}
+	spans, err := spansForAllRestoreTableIndexes(
+		backupCodec,
+		tables,
+		nil,   /* revs */
+		false, /* schemaOnly */
+		false, /* forOnlineRestore */
+	)
+	if err != nil {
+		return nil, err
+	}
+	completedSpans, completedIntroducedSpans, err := getCompletedSpans(
```

Comment:
just noting for the checkpoint PR: i hope we're checkpointing introduced spans correctly.